### PR TITLE
#9690 Implement a more conventional Docker setup

### PR DESCRIPTION
**Pull request by victoriachan**
# Description

Ticket: https://github.com/MozillaFoundation/foundation.mozilla.org/issues/9690

Move to a more conventional docker set up for local development.

# Checklist

**Tasks:**

- [x] Move to single mult stage Dockerfile
- [x] `inv new-env` builds a docker container locally and does not throw error.  
- [x] Running `docker-compose up` will make the website available at http://localhost:8000/cms/
- [x] New `inv start` command
- [x] CSS changes are reflected on the website (i.e. `npm run watch` is running)
- [x] New `inv start-lean` command for faster start up. This will use existing frontend assets instead of having to rebuild it. This is because npm run watch will always tear down and rebuild assets to ensure assets are up-to-date. This option will be useful for backend development where frontend assets are not changed.
- [x] Tasks lists in tasks.py still works (the frontend ones may need tweaking).
- [x] Test `invoke npm "install [PACKAGE]"` and `invoke npm update` still works and persist between inv start.
- [x] Fix `fatal: detected dubious ownership in repository at '/app'` issue when running `inv isort-check`.
- [x] Check that instructions for running playright locally still applies: https://github.com/MozillaFoundation/foundation.mozilla.org/blob/main/README.md#integration-tests
- [x] Test Visual Studio works as before
- [x] Update docker ignore
- [x] Unused files / code are removed. (e.g. old dockerfiles)
- [x] Documentation is updated

**GOALS (copied from ticket):**

- [x] The `inv new-env` command builds the Docker containers.
- [x] There is an `inv start` command to start the Docker stack. 
- [x] The `inv catchup` command rebuilds the frontend and backend containers. 
- [x] The database is not wiped during `inv catchup`.
- [x] The database is wiped during `inv new-env`.
- [x] There are no empty `dockerpythonenv` and `node_modules` directories in the project directory on the host? Maybe these need to exist (at least the `node_modules` directory might)...
- [x] When starting the container stack the frontend assets are already build and immediately available (based on the state when the container was last build). **This is the most important goal**, I think.
- [x] When the frontend container starts it runs a watch process to rebuild the frontend assets when the source files change. 
- [x] There is only one `Dockerfile` that builds the frontend and backend container as different stages. 
- [x] Even if we keep everything the same (which I don't think), it would be good to expand the documentation on our Docker setup.

**Tests**
- [x] Is the code I'm adding covered by tests?

**Changes in Models:**
- [x] Did I update or add new fake data?
- [x] Did I squash my migration?
- [x] Are my changes [backward-compatible](https://github.com/mozilla/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations). If not, did I schedule a deploy with the rest of the team?

**Documentation:**
- [x] Is my code documented?
- [x] Did I update the READMEs or wagtail documentation?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -64,7 +65,7 @@
     "tailwind": "run-p tailwind:watch sync",
     "tailwind:watch": "npm run build:tailwind -- -w",
     "test": "npm run lint",
-    "watch": "wait-on http://backend:8000/cms && run-s build:clean && run-p build:js:dev build:common watch:**",
+    "watch": "run-s build:clean && run-p build:js:dev build:common watch:**",
```
Comment (victoriachan) [inline package.json:68]: The frontend build was rather slow, so I removed the wait-on to speed things up a bit.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+ENV VIRTUAL_ENV=/app/dockerpythonvenv
+
+RUN useradd mozilla --create-home && mkdir /app $VIRTUAL_ENV && chown -R mozilla /app $VIRTUAL_ENV
+
+WORKDIR /app
+
+# Set default environment variables. They are used at build time and runtime.
+# If you specify your own environment variables on Heroku or Dokku, they will
+# override the ones set here. The ones below serve as sane defaults only.
+#  * PATH - Make sure that Poetry is on the PATH, along with our venv
```
Comment (jhonatan-lopes) [inline Dockerfile:41]: We might want to remove "Poetry" here, as we are not using it on Moz. Dependencies are managed through `pip-tools`.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+# Compile static files from static source at ./source to ./network-api/networkapi/frontend
+# This will create a `network-api/networkapi/frontend` directory.
+COPY ./source/ ./source/
+COPY ./network-api/networkapi/ ./network-api/networkapi/
+RUN npm run build
+
+
+# We use Debian images because they are considered more stable than the alpine
+# ones becase they use a different C compiler. Debian images also come with
```
Comment (jhonatan-lopes) [inline Dockerfile:23]: "becase" -> "because" (nitpick) :)

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+# all useful packages required for image manipulation out of the box. They
+# however weight a lot, approx. up to 1.5GiB per built image.
+#
+# Note: This stage builds the base image for production. Presently we are not
+# using this on the production site, but only use it as base for the dev build.
+FROM python:3.9.9-slim as base
```
Comment (jhonatan-lopes) [inline Dockerfile:29]: Loving the multistage build! :+1:

File: Dockerfile
Comment (jhonatan-lopes) [inline Dockerfile:1]: Do we need to reference **Dokku** in this file? I don't think we use it on Moz. If we don't, can we remove the references here?

File: Procfile.dev
Comment (jhonatan-lopes) [inline Procfile.dev:1]: Noob question: will this affect Heroku deployment?

File: docker-compose.yml [hunk 5]
```diff
@@ -1,26 +1,74 @@
 version: '3'
 
 services:
-
-  watch-static-files:
+  backend:
+    platform: linux/amd64 # Fix issues for M1 Macs
```
Comment (jhonatan-lopes) [inline docker-compose.yml:5]: That's awesome! :+1:

**Review by jhonatan-lopes — COMMENTED** [review]
Looking great, thanks @victoriachan! I have some minor comments, though

**Review by nickmoreton — COMMENTED** [review]
This all works really well overall.

One part that isn't working for me is the preview panel in the admin.

**Comment by jhonatan-lopes** [comment]
> This all works really well overall.
> 
> One part that isn't working for me is the preview panel in the admin.

If I recall correctly, I think that is tied to [this env var](https://github.com/MozillaFoundation/foundation.mozilla.org/blob/main/env.default#L61):

```bash
CSP_FRAME_ANCESTORS=" 'self' "
```

File: Procfile.dev
Comment (jhonatan-lopes) [reply Procfile.dev:1]: Great, thanks

File: Dockerfile (hunk 2 above)
Comment (victoriachan) [reply Dockerfile]: Ah you got me. I copied that from Wagtail Kit. Will remove.

File: Dockerfile (hunk 3 above)
Comment (victoriachan) [reply Dockerfile]: Corrected!

File: docs/local_development.md
[the remark this answers is not in the document]
Comment (victoriachan) [reply docs/local_development.md:1]: Good idea. That caught me out too. I've now added it to the README.

**Comment by victoriachan** [comment]
@nickmoreton @jhonatan-lopes Thank you for the review! I have addressed the issues you raised. However I haven't fixed the CSP issue on the Wagtail admin preview, as it looks like it is the same on the previous docker setup too, and not introduced by my change. Is there already a ticket for this elsewhere? It would be better to split it up to a separate ticket, would you agree?

**Comment by cdanfon** [comment]
Hey @victoriachan can you update the ticket description please? 

Currently this tasks is unticked but form our conversation I understand you've done it?:

- [ ] Did I update the READMEs or wagtail documentation?

**Comment by victoriachan** [comment]
Sorry, yes, I have done it. Have checked it now.

**Comment by jhonatan-lopes** [comment]
Hi @victoriachan, don't worry about the wagtail preview, this was already this way before this PR :)

**Comment by mmmavis** [comment]
Thanks for the work! 

I tested this out by cloning the branch to a new directory. `inv new-env` and `docker-compose up` both worked as expected.

**Comment by mmmavis** [comment]
I have a general Docker question and am wondering if any of you know a solution - 
What's the proper to colourize Docker logs? It's challenging to spot error/warning messages without colours. 👀

**Review by jhonatan-lopes — APPROVED** [review]

**Comment by mmmavis** [comment]
Percy doesn't seem to kick off for branches outside of Moz org. Devs have agreed to bypass Percy tests for this PR. If there's any issue after merging this in, we can always revert the changes and investigate the problem.

**Pull request merged**